Agent hangs during a connection attempt and delays shutdown

## Description  
When the agent is configured with an incorrect server address (`server_url`), it becomes unresponsive to shutdown signals (e.g., Ctrl+C). This behavior causes delays in stopping the agent and may be related to the retry mechanism for failed connection attempts.  

The issue appears to be influenced by the `retry_interval` setting, as indicated by the following log entry:  

```
[2024-11-27 16:47:13.017] [wazuh-agent] [warning] [WARN] [https_socket.hpp:56] [AsyncConnect] boost::asio::async_connect returned error code: 113 No route to host
[2024-11-27 16:47:13.018] [wazuh-agent] [warning] [WARN] [http_client.cpp:102] [Co_PerformHttpRequest] Failed to send http request. /api/v1/events/stateless. Retrying in 5 seconds.
^C
[2024-11-27 16:47:16.107] [wazuh-agent] [info] [INFO] [inventory.cpp:57] [Stop] Module stopped.
[2024-11-27 16:47:16.107] [wazuh-agent] [info] [INFO] [logcollector.cpp:51] [Stop] Logcollector stopped
[2024-11-27 16:47:16.123] [wazuh-agent] [info] [INFO] [inventory.cpp:36] [Start] Module finished.
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C
[2024-11-27 16:47:23.328] [wazuh-agent] [error] [ERROR] [https_socket.hpp:38] [Connect] Exception thrown: connect: No route to host [system:113 at /root/wazuh-agent/build/vcpkg_installed/x64-linux/include/boost/asio/detail/reactive_socket_service.hpp:587:33 in function 'boost::system::error_code boost::asio::detail::reactive_socket_service<boost::asio::ip::tcp>::connect(implementation_type &, const endpoint_type &, boost::system::error_code &) [Protocol = boost::asio::ip::tcp]']
[2024-11-27 16:47:23.329] [wazuh-agent] [error] [ERROR] [https_socket.hpp:79] [Write] Exception thrown during write: uninitialized (SSL routines) [asio.ssl:167772436]
[2024-11-27 16:47:23.329] [wazuh-agent] [error] [ERROR] [https_socket.hpp:112] [Read] Exception thrown during read: uninitialized (SSL routines) [asio.ssl:167772436]
[2024-11-27 16:47:23.329] [wazuh-agent] [error] [ERROR] [http_client.cpp:227] [AuthenticateWithUuidAndKey] Error parsing token in response: [json.exception.parse_error.101] parse error at line 1, column 1: attempting to parse an empty input; check that your input string or stream contains the expected JSON.
[2024-11-27 16:47:23.329] [wazuh-agent] [warning] [WARN] [communicator.cpp:32] [SendAuthenticationRequest] Failed to authenticate with the manager. Retrying in 5 seconds.
```  

## Steps to Reproduce  

1. Configure the agent with an incorrect `server_url`.  
2. Start the agent.  
3. Attempt to stop the agent using Ctrl+C.  

## Observed Behavior  

- The agent does not respond promptly to the shutdown signal.  
- The delay seems correlated with the connection retry interval.  

## Expected Behavior  

- The agent should respond immediately to the shutdown signal, regardless of its connection state or retry mechanism.

--- a/src/modules/inventory/include/inventory.hpp
+++ b/src/modules/inventory/include/inventory.hpp
@@ -74,7 +74,7 @@ class Inventory {
         void ScanPorts();
         void ScanProcesses();
         void Scan();
-        void SyncLoop(std::unique_lock<std::mutex>& lock);
+        void SyncLoop();
         void ShowConfig();
         cJSON * Dump() const;
         static void LogErrorInventory(const std::string& log);

--- a/src/modules/inventory/src/inventoryImp.cpp
+++ b/src/modules/inventory/src/inventoryImp.cpp
@@ -352,6 +352,8 @@ void Inventory::UpdateChanges(const std::string& table,
             NotifyChange(result, data, table);
         }
     };
+
+    std::unique_lock<std::mutex> lock{m_mutex};
     DBSyncTxn txn
     {
         m_spDBSync->handle(),
@@ -373,12 +375,16 @@ void Inventory::TryCatchTask(const std::function<void()>& task) const
     {
         if (!m_stopping)
         {
-            task();  // Ejecuta la tarea
+            task();
+        }
+        else
+        {
+            LogTrace("No Scanning during stopping");
         }
     }
     catch (const std::exception& ex)
     {
-        LogErrorInventory(std::string{ex.what()});
+        LogError("{}",std::string{ex.what()});
     }
 }
 
@@ -422,15 +428,18 @@ void Inventory::Init(const std::shared_ptr<ISysInfo>& spInfo,
     m_spInfo = spInfo;
     m_reportDiffFunction = reportDiffFunction;
 
-    std::unique_lock<std::mutex> lock{m_mutex};
-    m_stopping = false;
-    m_spDBSync = std::make_unique<DBSync>(HostType::AGENT,
-                                            DbEngineType::SQLITE3,
-                                            dbPath,
-                                            GetCreateStatement(),
-                                            DbManagement::PERSISTENT);
-    m_spNormalizer = std::make_unique<InvNormalizer>(normalizerConfigPath, normalizerType);
-    SyncLoop(lock);
+    {
+        std::unique_lock<std::mutex> lock{m_mutex};
+        m_stopping = false;
+        m_spDBSync = std::make_unique<DBSync>(HostType::AGENT,
+                                                DbEngineType::SQLITE3,
+                                                dbPath,
+                                                GetCreateStatement(),
+                                                DbManagement::PERSISTENT);
+        m_spNormalizer = std::make_unique<InvNormalizer>(normalizerConfigPath, normalizerType);
+    }
+
+    SyncLoop();
 }
 
 void Inventory::Destroy()
@@ -438,7 +447,6 @@ void Inventory::Destroy()
     std::unique_lock<std::mutex> lock{m_mutex};
     m_stopping = true;
     m_cv.notify_all();
-    lock.unlock();
 }
 
 
@@ -671,7 +679,6 @@ nlohmann::json Inventory::GetNetworkData()
                         networkTableData["network_item_id"] = GetItemId(networkTableData, NETWORK_ITEM_ID_FIELDS);
 
                         ret[NETWORKS_TABLE].push_back(networkTableData);
-
                     }
                 }
 
@@ -733,6 +740,8 @@ void Inventory::ScanPackages()
                 NotifyChange(result, data, PACKAGES_TABLE);
             }
         };
+
+        std::unique_lock<std::mutex> lock{m_mutex};
         DBSyncTxn txn
         {
             m_spDBSync->handle(),
@@ -862,6 +871,7 @@ void Inventory::ScanProcesses()
                 NotifyChange(result, data, PROCESSES_TABLE);
             }
         };
+        std::unique_lock<std::mutex> lock{m_mutex};
         DBSyncTxn txn
         {
             m_spDBSync->handle(),
@@ -902,19 +912,24 @@ void Inventory::Scan()
     LogInfo("Evaluation finished.");
 }
 
-void Inventory::SyncLoop(std::unique_lock<std::mutex>& lock)
+void Inventory::SyncLoop()
 {
-    if (m_scanOnStart)
+    LogInfo("Module started.");
+
+    if (m_scanOnStart && !m_stopping)
     {
         Scan();
     }
 
-    while (!m_cv.wait_for(lock, std::chrono::milliseconds{m_intervalValue}, [&]()
-{
-    return m_stopping;
-}))
+    while (!m_stopping)
     {
+        {
+            std::unique_lock<std::mutex> lock{m_mutex};
+            m_cv.wait_for(lock,
+                std::chrono::milliseconds{m_intervalValue}, [&]() { return m_stopping; } );
+        }
         Scan();
     }
+    std::unique_lock<std::mutex> lock{m_mutex};
     m_spDBSync.reset(nullptr);
 }

